File: claudeshome/lua/wsserver.lua
--[[
  wsserver.lua — Non-blocking WebSocket server for Love2D

  Accepts multiple client connections, performs WebSocket handshakes,
  reads masked frames from clients, sends unmasked frames to clients.

  Uses love.data.hash("sha1") and love.data.encode("base64") for
  the Sec-WebSocket-Accept handshake computation.

  Usage:
    local wsserver = require("lua.wsserver")
    local server = wsserver.new("127.0.0.1", 8080)

    function server:onconnect(clientId) end
    function server:onmessage(clientId, msg) end
    function server:ondisconnect(clientId) end

    -- Each frame:
    server:update()
    server:broadcast("hello everyone")
    server:send(clientId, "hello you")

    -- Shutdown:
    server:close()
]]

local socket = require("socket")
local bit = require("bit")
local band, bor, bxor = bit.band, bit.bor, bit.bxor
local shl, shr = bit.lshift, bit.rshift

local MAGIC_GUID = "258EAFA5-E914-47DA-95CA-5AB515859764"

local OPCODE = {
  CONTINUE = 0,
  TEXT     = 1,
  BINARY   = 2,
  CLOSE    = 8,
  PING     = 9,
  PONG     = 10,
}

local _M = {}
_M.__index = _M

-- Default callbacks (override on instance)
function _M:onconnect(clientId) end
function _M:onmessage(clientId, message) end
function _M:ondisconnect(clientId) end

--- Compute Sec-WebSocket-Accept from client's Sec-WebSocket-Key.
local function computeAcceptKey(clientKey)
  local hash = love.data.hash("sha1", clientKey .. MAGIC_GUID)
  return love.data.encode("string", "base64", hash)
end

--- Send an unmasked WebSocket frame (server → client).
local function sendFrame(sock, opcode, message)
  -- FIN + opcode
  local ok, err = sock:send(string.char(bor(0x80, opcode)))
  if not ok then return nil, err end

  if not message or #message == 0 then
    return sock:send(string.char(0))
  end

  local length = #message
  if length > 65535 then
    sock:send(string.char(127,
      0, 0, 0, 0,
      band(shr(length, 24), 0xff),
      band(shr(length, 16), 0xff),
      band(shr(length, 8), 0xff),
      band(length, 0xff)))
  elseif length > 125 then
    sock:send(string.char(126,
      band(shr(length, 8), 0xff),
      band(length, 0xff)))
  else
    sock:send(string.char(length))
  end

  return sock:send(message)
end

--- Create a new WebSocket server.
--- @param host string  Bind address (e.g., "127.0.0.1")
--- @param port number  Listen port
--- @return table server
function _M.new(host, port)
  local serverSock = socket.tcp()
  serverSock:settimeout(0)
  serverSock:setoption("reuseaddr", true)

  local ok, err = serverSock:bind(host or "127.0.0.1", port)
  if not ok then
    return nil, "Failed to bind: " .. tostring(err)
  end

  serverSock:listen(16)

  local m = {
    socket = serverSock,
    host = host or "127.0.0.1",
    port = port,
    clients = {},       -- clientId -> { sock, state, buffer, ... }
    nextClientId = 1,
  }

  setmetatable(m, _M)
  io.write("[wsserver] Listening on " .. (host or "127.0.0.1") .. ":" .. port .. "\n"); io.flush()
  return m
end

--- Accept pending connections and process all clients.
--- Call once per frame.
function _M:update()
  -- Accept new TCP connections
  while true do
    local client, err = self.socket:accept()
    if not client then break end

    client:settimeout(0)
    local id = self.nextClientId
    self.nextClientId = id + 1

    self.clients[id] = {
      sock = client,
      state = "handshake",  -- "handshake" → "open" → "closed"
      buffer = "",
      frameBuffer = "",
      frameLength = 0,
      frameHead = nil,
      continue = "",
    }
  end

  -- Process each client
  for id, c in pairs(self.clients) do
    if c.state == "handshake" then
      self:_processHandshake(id, c)
    elseif c.state == "open" then
      self:_processFrames(id, c)
    end
  end
end

--- Process WebSocket handshake for a client.
function _M:_processHandshake(id, c)
  -- Drain all available lines from the socket
  while true do
    local data, err, partial = c.sock:receive("*l")
    if data then
      -- Full line received (LuaSocket strips the trailing \n)
      c.buffer = c.buffer .. data .. "\r\n"
      -- Empty line signals end of HTTP headers
      if data == "" or data == "\r" then break end
    elseif partial and #partial > 0 then
      -- Incomplete line — append what we have and wait for more
      c.buffer = c.buffer .. partial
      return
    else
      -- No data available
      if err == "closed" then
        self:_removeClient(id)
      end
      return
    end
  end

  -- Check if we have the full headers (double CRLF)
  if not (c.buffer:find("\r\n\r\n") or c.buffer:find("\n\n")) then
    return  -- Need more data
  end

  -- Extract Sec-WebSocket-Key
  local key = c.buffer:match("[Ss]ec%-[Ww]eb[Ss]ocket%-[Kk]ey:%s*([^\r\n]+)")
  if not key then
    -- Not a valid WebSocket upgrade request
    c.sock:send("HTTP/1.1 400 Bad Request\r\n\r\n")
    self:_removeClient(id)
    return
  end

  key = key:match("^%s*(.-)%s*$")  -- Trim

  -- Send upgrade response
  local acceptKey = computeAcceptKey(key)
  local response = "HTTP/1.1 101 Switching Protocols\r\n" ..
    "Upgrade: websocket\r\n" ..
    "Connection: Upgrade\r\n" ..
    "Sec-WebSocket-Accept: " .. acceptKey .. "\r\n\r\n"

  local ok, err = c.sock:send(response)
  if not ok then
    self:_removeClient(id)
    return
  end

  c.state = "open"
  c.buffer = ""
  c.frameBuffer = ""
  c.frameLength = 2  -- Start reading 2-byte frame header
  c.frameHead = nil

  self:onconnect(id)
end

--- Read and process WebSocket frames from a client.
--- Handles masked frames (client → server).
function _M:_processFrames(id, c)
  while true do
    -- Try to read remaining bytes for current frame
    local needed = c.frameLength - #c.frameBuffer
    if needed > 0 then
      local data, err, partial = c.sock:receive(needed)
      if err == "closed" then
        self:_removeClient(id, 1006, "connection lost")
        return
      end
      if data then
        c.frameBuffer = c.frameBuffer .. data
      elseif partial and #partial > 0 then
        c.frameBuffer = c.frameBuffer .. partial
      else
        return  -- No data available
      end
    end

    if #c.frameBuffer < c.frameLength then
      return  -- Still need more data
    end

    -- Parse frame header
    if not c.frameHead then
      if #c.frameBuffer < 2 then return end

      local b1, b2 = c.frameBuffer:byte(1, 2)
      c.frameHead = b1
      local masked = band(b2, 0x80) == 0x80
      local length = band(b2, 0x7f)

      if length == 126 then
        -- Need 2 more bytes for extended length
        if c.frameLength == 2 then
          c.frameLength = 4
          goto continue
        end
        if #c.frameBuffer < 4 then return end
        local b3, b4 = c.frameBuffer:byte(3, 4)
        length = shl(b3, 8) + b4
      elseif length == 127 then
        -- Need 8 more bytes for extended length
        if c.frameLength == 2 then
          c.frameLength = 10
          goto continue
        end
        if #c.frameBuffer < 10 then return end
        local b7, b8, b9, b10 = c.frameBuffer:byte(7, 10)
        length = shl(b7, 24) + shl(b8, 16) + shl(b9, 8) + b10
      end

      -- Calculate total frame size: header consumed + mask(4 if masked) + payload
      local headerSize = #c.frameBuffer  -- bytes consumed so far for header
      local totalNeeded = headerSize + (masked and 4 or 0) + length
      c.frameLength = totalNeeded
      c._masked = masked
      c._payloadLength = length
      c._headerSize = headerSize

      if #c.frameBuffer < totalNeeded then
        goto continue
      end
    end

    -- Full frame received — extract payload
    local payload
    local offset = c._headerSize + 1  -- 1-based index after header

    if c._masked then
      -- Read 4-byte mask key
      local m1, m2, m3, m4 = c.frameBuffer:byte(offset, offset + 3)
      local maskKey = { m1, m2, m3, m4 }
      offset = offset + 4

      -- Unmask payload
      local payloadBytes = { c.frameBuffer:byte(offset, offset + c._payloadLength - 1) }
      for i = 1, #payloadBytes do
        payloadBytes[i] = bxor(payloadBytes[i], maskKey[(i - 1) % 4 + 1])
      end
      payload = string.char(unpack(payloadBytes))
    else
      payload = c.frameBuffer:sub(offset, offset + c._payloadLength - 1)
    end

    -- Dispatch by opcode
    local opcode = band(c.frameHead, 0x0f)
    local fin = band(c.frameHead, 0x80) == 0x80

    if opcode == OPCODE.CLOSE then
      -- Send close frame back
      pcall(sendFrame, c.sock, OPCODE.CLOSE, nil)
      self:_removeClient(id, 1000, "")
      return
    elseif opcode == OPCODE.PING then
      sendFrame(c.sock, OPCODE.PONG, payload)
    elseif opcode == OPCODE.CONTINUE then
      c.continue = c.continue .. payload
      if fin then
        self:onmessage(id, c.continue)
        c.continue = ""
      end
    else
      if fin then
        self:onmessage(id, payload)
      else
        c.continue = payload
      end
    end

    -- Reset for next frame
    c.frameBuffer = ""
    c.frameLength = 2
    c.frameHead = nil
    c._masked = nil
    c._payloadLength = nil
    c._headerSize = nil

    ::continue::
  end
end

--- Remove a client and fire ondisconnect.
function _M:_removeClient(id, code, reason)
  local c = self.clients[id]
  if not c then return end

  pcall(function() c.sock:close() end)
  self.clients[id] = nil
  self:ondisconnect(id, code or 1005, reason or "")
end

--- Send a text message to a specific client.
--- @param clientId number
--- @param message string
function _M:send(clientId, message)
  local c = self.clients[clientId]
  if not c or c.state ~= "open" then return end

  local ok, err = sendFrame(c.sock, OPCODE.TEXT, message)
  if not ok then
    self:_removeClient(clientId, 1006, "send failed")
  end
end

--- Broadcast a text message to all connected clients.
--- @param message string
function _M:broadcast(message)
  for id, c in pairs(self.clients) do
    if c.state == "open" then
      local ok, err = sendFrame(c.sock, OPCODE.TEXT, message)
      if not ok then
        self:_removeClient(id, 1006, "send failed")
      end
    end
  end
end

--- Get the number of connected clients.
--- @return number
function _M:clientCount()
  local count = 0
  for _, c in pairs(self.clients) do
    if c.state == "open" then count = count + 1 end
  end
  return count
end

--- Get list of connected client IDs.
--- @return number[]
function _M:getClientIds()
  local ids = {}
  for id, c in pairs(self.clients) do
    if c.state == "open" then ids[#ids + 1] = id end
  end
  return ids
end

--- Shut down the server and disconnect all clients.
function _M:close()
  for id, c in pairs(self.clients) do
    pcall(sendFrame, c.sock, OPCODE.CLOSE, nil)
    pcall(function() c.sock:close() end)
  end
  self.clients = {}

  pcall(function() self.socket:close() end)
  io.write("[wsserver] Closed\n"); io.flush()
end

return _M

File: tsz/runtime/compiled/net/wsserver.zig
//! WebSocket server — non-blocking, multi-client, broadcast + unicast.
//!
//! Port of love2d/lua/wsserver.lua. Accepts connections, performs HTTP
//! upgrade handshake, then frames flow. Non-blocking poll each frame.
//!
//! Usage:
//!   var server = try wsserver.listen(8080);
//!   // each frame:
//!   var events: [32]wsserver.ServerEvent = undefined;
//!   const n = server.update(&events);
//!   for (events[0..n]) |ev| { ... }
//!   server.broadcast("hello everyone");
//!   server.send(client_id, "hello you");
//!   server.close();

const std = @import("std");

// ── Constants ────────────────────────────────────────────────────────────

const MAGIC_GUID = "258EAFA5-E914-47DA-95CA-5AB515859764";
const MAX_CLIENTS = 64;
const MAX_MSG = 65536;
const MAX_HDR = 4096;

// ── Public types ─────────────────────────────────────────────────────────

pub const ServerEventType = enum {
    client_connected,
    client_message,
    client_disconnected,
};

pub const ServerEvent = struct {
    client_id: u32 = 0,
    event_type: ServerEventType = .client_connected,
    data: [MAX_MSG]u8 = undefined,
    data_len: usize = 0,

    pub fn dataSlice(self: *const ServerEvent) []const u8 {
        return self.data[0..self.data_len];
    }
};

const ClientStatus = enum { handshake, open, closed };

const Client = struct {
    active: bool = false,
    id: u32 = 0,
    stream: ?std.net.Stream = null,
    status: ClientStatus = .closed,
    handshake_buf: [MAX_HDR]u8 = undefined,
    handshake_len: usize = 0,
    read_buf: [MAX_MSG]u8 = undefined,
    read_len: usize = 0,
};

// ── Server ───────────────────────────────────────────────────────────────

pub const WsServer = struct {
    listener: std.posix.socket_t,
    clients: [MAX_CLIENTS]Client = [_]Client{.{}} ** MAX_CLIENTS,
    next_client_id: u32 = 1,
    event_buf: [64]ServerEvent = undefined,
    event_count: usize = 0,

    /// Start listening on a port.
    pub fn listen(port: u16) !WsServer {
        const addr = try std.net.Address.parseIp4("0.0.0.0", port);
        const fd = try std.posix.socket(addr.any.family, std.posix.SOCK.STREAM | std.posix.SOCK.NONBLOCK, 0);
        errdefer std.posix.close(fd);

        // SO_REUSEADDR
        const optval: c_int = 1;
        try std.posix.setsockopt(fd, std.posix.SOL.SOCKET, std.posix.SO.REUSEADDR, std.mem.asBytes(&optval));

        try std.posix.bind(fd, &addr.any, addr.getOsSockLen());
        try std.posix.listen(fd, 16);

        return WsServer{ .listener = fd };
    }

    /// Non-blocking poll. Call once per frame. Returns event count.
    pub fn update(self: *WsServer, out: []ServerEvent) usize {
        self.event_count = 0;

        // Accept new connections
        self.acceptClients(out);

        // Process existing clients
        for (&self.clients) |*client| {
            if (!client.active) continue;
            switch (client.status) {
                .handshake => self.processHandshake(client, out),
                .open => self.processFrames(client, out),
                .closed => {},
            }
        }

        return @min(self.event_count, out.len);
    }

    /// Send to a specific client.
    pub fn send(self: *WsServer, client_id: u32, data: []const u8) void {
        for (&self.clients) |*client| {
            if (client.active and client.id == client_id and client.status == .open) {
                if (client.stream) |stream| {
                    writeServerFrameWithOpcode(stream, 0x81, data) catch {
                        client.status = .closed;
                        client.active = false;
                    };
                }
                return;
            }
        }
    }

    /// Broadcast to all connected clients.
    pub fn broadcast(self: *WsServer, data: []const u8) void {
        for (&self.clients) |*client| {
            if (client.active and client.status == .open) {
                if (client.stream) |stream| {
                    writeServerFrameWithOpcode(stream, 0x81, data) catch {
                        client.status = .closed;
                        client.active = false;
                    };
                }
            }
        }
    }

    /// Shut down the server and all clients.
    pub fn close(self: *WsServer) void {
        for (&self.clients) |*client| {
            if (client.active) {
                if (client.stream) |s| s.close();
                client.active = false;
            }
        }
        std.posix.close(self.listener);
    }

    // ── Internal ─────────────────────────────────────────────────────

    fn acceptClients(self: *WsServer, out: []ServerEvent) void {
        // Try to accept (non-blocking)
        while (true) {
            const accepted = std.posix.accept(self.listener, null, null, std.posix.SOCK.NONBLOCK) catch break;
            const slot = self.findClientSlot() orelse {
                std.posix.close(accepted);
                break;
            };
            self.clients[slot] = .{
                .active = true,
                .id = self.next_client_id,
                .stream = .{ .handle = accepted },
                .status = .handshake,
            };
            self.next_client_id += 1;
            _ = out;
        }
    }

    fn processHandshake(self: *WsServer, client: *Client, out: []ServerEvent) void {
        const stream = client.stream orelse return;
        // Read HTTP upgrade request
        const n = stream.read(client.handshake_buf[client.handshake_len..]) catch |err| {
            if (err == error.WouldBlock) return;
            client.status = .closed;
            client.active = false;
            return;
        };
        if (n == 0) {
            client.status = .closed;
            client.active = false;
            return;
        }
        client.handshake_len += n;

        // Check for complete headers
        const headers = client.handshake_buf[0..client.handshake_len];
        if (std.mem.indexOf(u8, headers, "\r\n\r\n") == null) return;

        // Extract Sec-WebSocket-Key
        const key = extractHeader(headers, "Sec-WebSocket-Key: ") orelse {
            client.status = .closed;
            client.active = false;
            return;
        };

        // Compute accept hash: SHA1(key + MAGIC_GUID) base64
        var hasher = std.crypto.hash.Sha1.init(.{});
        hasher.update(key);
        hasher.update(MAGIC_GUID);
        const digest = hasher.finalResult();
        const accept = std.base64.standard.Encoder.encode(&([_]u8{0} ** 28), &digest);

        // Send 101 response
        const writer = stream.writer();
        writer.print(
            "HTTP/1.1 101 Switching Protocols\r\n" ++
                "Upgrade: websocket\r\n" ++
                "Connection: Upgrade\r\n" ++
                "Sec-WebSocket-Accept: {s}\r\n\r\n",
            .{accept},
        ) catch {
            client.status = .closed;
            client.active = false;
            return;
        };

        client.status = .open;
        self.pushEvent(out, client.id, .client_connected, "");
    }

    fn processFrames(self: *WsServer, client: *Client, out: []ServerEvent) void {
        const stream = client.stream orelse return;

        // Read available data
        if (client.read_len < MAX_MSG) {
            const n = stream.read(client.read_buf[client.read_len..]) catch |err| {
                if (err == error.WouldBlock) {
                    if (client.read_len >= 2) self.tryParseClientFrame(client, out);
                    return;
                }
                self.pushEvent(out, client.id, .client_disconnected, "");
                client.status = .closed;
                client.active = false;
                return;
            };
            if (n == 0) {
                self.pushEvent(out, client.id, .client_disconnected, "");
                client.status = .closed;
                client.active = false;
                return;
            }
            client.read_len += n;
        }

        self.tryParseClientFrame(client, out);
    }

    fn tryParseClientFrame(self: *WsServer, client: *Client, out: []ServerEvent) void {
        if (client.read_len < 2) return;

        const byte0 = client.read_buf[0];
        const byte1 = client.read_buf[1];
        const opcode = byte0 & 0x0F;
        const masked = (byte1 & 0x80) != 0;
        var payload_len: usize = byte1 & 0x7F;
        var header_len: usize = 2;

        if (payload_len == 126) {
            if (client.read_len < 4) return;
            payload_len = (@as(usize, client.read_buf[2]) << 8) | client.read_buf[3];
            header_len = 4;
        } else if (payload_len == 127) {
            if (client.read_len < 10) return;
            payload_len = 0;
            for (2..10) |i| {
                payload_len = (payload_len << 8) | client.read_buf[i];
            }
            header_len = 10;
        }
        payload_len = @min(payload_len, MAX_MSG);

        var mask_key: [4]u8 = .{ 0, 0, 0, 0 };
        if (masked) {
            if (client.read_len < header_len + 4) return;
            @memcpy(&mask_key, client.read_buf[header_len .. header_len + 4]);
            header_len += 4;
        }

        const total = header_len + payload_len;
        if (client.read_len < total) return;

        // Extract and unmask payload
        var payload: [MAX_MSG]u8 = undefined;
        @memcpy(payload[0..payload_len], client.read_buf[header_len .. header_len + payload_len]);
        if (masked) {
            for (0..payload_len) |i| payload[i] ^= mask_key[i % 4];
        }

        // Consume from buffer
        const remaining = client.read_len - total;
        if (remaining > 0) {
            std.mem.copyForwards(u8, client.read_buf[0..remaining], client.read_buf[total..client.read_len]);
        }
        client.read_len = remaining;

        // Handle opcode
        if (opcode == 1 or opcode == 2) {
            // Text or binary
            self.pushEvent(out, client.id, .client_message, payload[0..payload_len]);
        } else if (opcode == 8) {
            // Close
            self.pushEvent(out, client.id, .client_disconnected, "");
            if (client.stream) |s| s.close();
            client.status = .closed;
            client.active = false;
        } else if (opcode == 9) {
            // Ping → send pong (opcode 0x0A)
            if (client.stream) |s| writeServerFrameWithOpcode(s, 0x8A, payload[0..payload_len]) catch {};
        }
    }

    fn findClientSlot(self: *WsServer) ?usize {
        for (0..MAX_CLIENTS) |i| {
            if (!self.clients[i].active) return i;
        }
        return null;
    }

    fn pushEvent(self: *WsServer, out: []ServerEvent, client_id: u32, event_type: ServerEventType, data: []const u8) void {
        if (self.event_count >= out.len) return;
        var ev = &out[self.event_count];
        ev.client_id = client_id;
        ev.event_type = event_type;
        const dlen = @min(data.len, MAX_MSG);
        if (dlen > 0) @memcpy(ev.data[0..dlen], data[0..dlen]);
        ev.data_len = dlen;
        self.event_count += 1;
    }
};

// ── Helpers ──────────────────────────────────────────────────────────────

fn extractHeader(headers: []const u8, name: []const u8) ?[]const u8 {
    var pos: usize = 0;
    while (pos < headers.len) {
        if (std.mem.startsWith(u8, headers[pos..], name)) {
            const start = pos + name.len;
            const end = std.mem.indexOf(u8, headers[start..], "\r\n") orelse (headers.len - start);
            return headers[start .. start + end];
        }
        // Skip to next line
        if (std.mem.indexOf(u8, headers[pos..], "\r\n")) |nl| {
            pos += nl + 2;
        } else break;
    }
    return null;
}

/// Write an unmasked server→client frame with a specific first byte (FIN+opcode).
fn writeServerFrameWithOpcode(stream: std.net.Stream, first_byte: u8, payload: []const u8) !void {
    const writer = stream.writer();
    try writer.writeByte(first_byte);
    if (payload.len > 65535) {
        try writer.writeByte(127);
        var len_bytes: [8]u8 = undefined;
        std.mem.writeInt(u64, &len_bytes, payload.len, .big);
        try writer.writeAll(&len_bytes);
    } else if (payload.len > 125) {
        try writer.writeByte(126);
        var len_bytes: [2]u8 = undefined;
        std.mem.writeInt(u16, &len_bytes, @intCast(payload.len), .big);
        try writer.writeAll(&len_bytes);
    } else {
        try writer.writeByte(@intCast(payload.len));
    }
    try writer.writeAll(payload);
}
